src/log/Entry.h

```diff
@@ -4,102 +4,68 @@
-  Entry *m_next;
+#include <pthread.h>
 
-  size_t m_buf_len;
-  size_t* m_exp_len;
-  char m_static_buf[1];
+#include <string_view>
 
-  prebuffered_data m_data;
-  CachedPrebufferedStreambuf* m_streambuf;
+namespace ceph {
+namespace logging {
 
-  Entry()
-    : Entry(log_time{}, 0, 0, 0, nullptr)
+class Entry {
+public:
+  using time = log_time;
+
+  Entry() = default;
+  Entry(short pr, short sub) :
+    m_stamp(clock().now()),
+    m_thread(pthread_self()),
+    m_prio(pr),
+    m_subsys(sub)
   {}
-  Entry(log_time s, pthread_t t, short pr, short sub,
-	const char *msg = NULL)
-    : Entry(s, t, pr, sub, m_static_buf, sizeof(m_static_buf), nullptr,
-	    msg)
+  Entry(const Entry &) = delete;
+  Entry& operator=(const Entry &) = delete;
+  Entry(Entry &&e) :
+    m_stamp(e.m_stamp),
+    m_thread(e.m_thread),
+    m_prio(e.m_prio),
+    m_subsys(e.m_subsys),
+    os(std::move(e.os))
   {}
-  Entry(log_time s, pthread_t t, short pr, short sub, char* buf, size_t buf_len, size_t* exp_len,
-	const char *msg = NULL)
-    : m_stamp(s), m_thread(t), m_prio(pr), m_subsys(sub),
-      m_next(NULL),
-      m_buf_len(buf_len),
-      m_exp_len(exp_len),
-      m_data(buf, buf_len),
-      m_streambuf(CachedPrebufferedStreambuf::create(&m_data))
-  {
-    if (msg) {
-      get_ostream() << msg;
-    }
+  Entry& operator=(Entry &&e) {
+    m_stamp = e.m_stamp;
+    m_thread = e.m_thread;
+    m_prio = e.m_prio;
+    m_subsys = e.m_subsys;
+    os = std::move(e.os);
+    return *this;
   }
+  ~Entry() {}
 
-private:
-  ~Entry() = default;
-
-public:
-  std::ostream& get_ostream() {
-    return m_streambuf->get_ostream();
+  std::string_view strv() const {
+    return os.strv();
   }
-
-  // function improves estimate for expected size of message
-  void hint_size() {
-    if (m_exp_len != NULL) {
-      size_t size = m_data.size();
-      if (size > __atomic_load_n(m_exp_len, __ATOMIC_RELAXED)) {
-        //log larger then expected, just expand
-        __atomic_store_n(m_exp_len, size + 10, __ATOMIC_RELAXED);
-      }
-      else {
-        //asymptotically adapt expected size to message size
-        __atomic_store_n(m_exp_len, (size + 10 + m_buf_len*31) / 32, __ATOMIC_RELAXED);
-      }
-    }
-  }
-
-  void set_str(const std::string &s) {
-    get_ostream() << s;
-  }
-
-  std::string get_str() const {
-    return m_data.get_str();
-  }
-
-  // returns current size of content
   size_t size() const {
-    return m_data.size();
+    return os.strv().size();
   }
 
-  // extracts up to avail chars of content
-  int snprintf(char* dst, size_t avail) const {
-    return m_data.snprintf(dst, avail);
-  }
+  time m_stamp;
+  pthread_t m_thread;
+  short m_prio, m_subsys;
 
-  void finish() {
-    m_streambuf->finish();
+  std::ostream& get_ostream() {
+    return os;
   }
 
-  void destroy() {
-    if (m_exp_len != NULL) {
-      this->~Entry();
-      ::operator delete(this);
-    } else {
-      delete(this);
-    }
+private:
+  StackStringStream os;
```

Comment:
before this change, the `streambuf` used by log entries is cached on a per-thread basis, and each thread has its own `streambuf`, so log entries created on a given thread do not need to create and destroy `streambuf` for stringifying the message to print. 

after this change, the `streambuf` is created and destroyed on the stack for every single log `Entry`.

> basic_string::_M_mutate is also another allocation via PrebufferedStreambuf.

i think what you meant is `CachedPrebufferedStreambuf`. `PrebufferedStreambuf`  is not used anymore. as i explained above, the creation and destructure of `CachedPrebufferedStreambuf` is an one-shot operation on each thread. so i think the overhead introduced by it is fine. and it will be offset by the reuse of the cached streambuf.